File: apps/grokking-ml-kotlin-lib/src/main/kotlin/lib/Nb13NeuralNetworkBoundary.kt
package lib

import kotlin.math.ceil
import org.jetbrains.kotlinx.dataframe.DataFrame
import org.jetbrains.kotlinx.dataframe.api.convertTo
import org.jetbrains.kotlinx.dataframe.api.getColumn
import org.jetbrains.kotlinx.dataframe.io.readCSV
import smile.base.mlp.Layer
import smile.base.mlp.OutputFunction
import smile.classification.MLP

/**
 * 原著ノートブック #13 `Chapter_10_Neural_Networks/Graphical_example.ipynb`。
 *
 * ニューラルネットワークの章に入る。円形に分布した 110 点を、
 * **2 層の隠れ層を持つネットワーク** で分類し、決定境界を見る。
 *
 * 決定木（[#10][Nb10DecisionTreeBoundary]）の境界が軸に平行な長方形だったのに対し、
 * ニューラルネットワークの境界は **曲線** になる。
 *
 * 原著は Keras で `Dense(128, relu)` → `Dropout` → `Dense(64, relu)` → `Dropout`
 * → `Dense(2, softmax)` と組む。Smile の [MLP] には **Dropout が無い** ので、
 * 隠れ層 2 つだけの構成になる。
 */
object Nb13NeuralNetworkBoundary {

    /** 原著のネットワークの隠れ層 */
    val HIDDEN_UNITS = intArrayOf(128, 64)

    /**
     * 原著の学習回数は 100 だが、Smile では足りない。
     *
     * Keras は Adam（学習率を自動調整する最適化）を使うのに対し、Smile の [MLP] は
     * 単純な確率的勾配降下である。実測すると 100 回では多数派を答えるだけの
     * 正解率 0.764 にとどまり、2000 回で 0.891 に届いた。
     */
    const val EPOCHS = 2000

    /** 原著が指定している学習回数。比較用に残す */
    const val ORIGINAL_EPOCHS = 100

    /** 原著の `plot_model` が使う格子の刻み幅 */
    const val PLOT_STEP = 0.2

    /** 読み込んだ円形データ */
    class Circle(val x: Array<DoubleArray>, val y: IntArray) {
        val size: Int get() = y.size
    }

    /** 円形に分布したデータを読み込む。110 点 */
    fun loadCircle(): Circle {
        val frame: DataFrame<*> = DataFrame.readCSV(Datasets.path("one_circle.csv").toFile())
        val x1 = frame.getColumn("x_1").convertTo<Double>().toList()
        val x2 = frame.getColumn("x_2").convertTo<Double>().toList()
        val y = frame.getColumn("y").convertTo<Int>().toList()
        return Circle(Array(x1.size) { doubleArrayOf(x1[it], x2[it]) }, y.toIntArray())
    }

    /**
     * 原著と同じ隠れ層を持つネットワークを学習する。
     *
     * Smile の [MLP] は層を [Layer] のビルダーで組む。
     * `Layer.input` が入力層、`Layer.rectifier` が ReLU、
     * `Layer.mle` が出力層である。
     *
     * **Dropout は無い。** 原著は過学習を抑えるために 2 か所へ入れているが、
     * Smile では代わりに学習回数で調整することになる。
     *
     * 出力は 2 クラスなので `OutputFunction.SIGMOID` を使う。
     * Keras の `softmax` は 2 クラスならシグモイドと等価な境界を与える。
     */
    fun fit(data: Circle, epochs: Int = EPOCHS, seed: Int = 0): MLP {
        // Smile の MLP は内部で乱数を使うので、種を固定して再現性を確保する
        smile.math.MathEx.setSeed(seed.toLong())

        val network = MLP(
            Layer.input(2),
            Layer.rectifier(HIDDEN_UNITS[0]),
            Layer.rectifier(HIDDEN_UNITS[1]),
            Layer.mle(1, OutputFunction.SIGMOID),
        )
        repeat(epochs) { network.update(data.x, data.y) }
        return network
    }

    /** 学習データに対する正解率 */
    fun accuracy(model: MLP, data: Circle): Double {
        val correct = data.x.indices.count { model.predict(data.x[it]) == data.y[it] }
        return correct.toDouble() / data.size
    }

    /** NumPy の `np.arange` と同じく、終端を含まない等差数列を作る */
    fun arange(start: Double, stop: Double, step: Double): DoubleArray {
        val count = ceil((stop - start) / step).toInt()
        return DoubleArray(count) { start + it * step }
    }

    /** 原著の `plot_model` と同じ格子を作り、各点の予測クラスを返す */
    fun decisionGrid(model: MLP, data: Circle, step: Double = PLOT_STEP): Array<IntArray> {
        val xs = data.x.map { it[0] }
        val ys = data.x.map { it[1] }
        val xValues = arange(xs.min() - 1, xs.max() + 1, step)
        val yValues = arange(ys.min() - 1, ys.max() + 1, step)

        return Array(yValues.size) { row ->
            IntArray(xValues.size) { column ->
                model.predict(doubleArrayOf(xValues[column], yValues[row]))
            }
        }
    }

    /**
     * 各行で予測が切り替わった回数を返す。
     *
     * 決定木なら軸に平行な境界なので、切り替わる位置は行によらず同じだった。
     * ニューラルネットワークは曲線を引けるので、行ごとに変わる。
     */
    fun boundaryChangesPerRow(grid: Array<IntArray>): List<Int> =
        grid.map { row -> (1 until row.size).count { row[it] != row[it - 1] } }
}

File: apps/grokking-ml-fsharp-lib/src/GrokkingMlLib/Nb13NeuralNetworkBoundary.fs
/// 原著ノートブック #13 `Chapter_10_Neural_Networks/Graphical_example.ipynb`。
///
/// ニューラルネットワークの章に入る。円形に分布した 110 点を、
/// **2 層の隠れ層を持つネットワーク** で分類し、決定境界を見る。
///
/// 決定木（[#10](Nb10DecisionTreeBoundary.fs)）の境界が軸に平行な長方形だったのに対し、
/// ニューラルネットワークの境界は **曲線** になる。
///
/// 原著は Keras で `Dense(128, relu)` → `Dropout` → `Dense(64, relu)` → `Dropout`
/// → `Dense(2, softmax)` と組む。Accord.Neuro には **Dropout も ReLU も無い** ので、
/// シグモイドの隠れ層 2 つで代用する。
module GrokkingMlLib.Nb13NeuralNetworkBoundary

open Accord.Neuro
open Accord.Neuro.Learning
open Deedle

/// 原著のネットワークの隠れ層
let hiddenUnits = [| 128; 64 |]

/// 原著の学習回数
[<Literal>]
let OriginalEpochs = 100

/// このネットワークで十分に学習が進む回数
[<Literal>]
let Epochs = 1000

/// 学習率。Accord の既定（0.1）をそのまま使う
[<Literal>]
let LearningRate = 0.1

/// 原著の `plot_model` が使う格子の刻み幅
[<Literal>]
let PlotStep = 0.2

/// 読み込んだ円形データ
type Circle =
    { X: float[][]
      Y: int[] }

    member this.Size = this.Y.Length

/// 円形に分布したデータを読み込む。110 点
let loadCircle () =
    let frame = Datasets.loadFrame "one_circle.csv"
    let x1 = frame.GetColumn<float>("x_1") |> Series.values |> Array.ofSeq
    let x2 = frame.GetColumn<float>("x_2") |> Series.values |> Array.ofSeq
    let y = frame.GetColumn<int>("y") |> Series.values |> Array.ofSeq

    { X = Array.map2 (fun a b -> [| a; b |]) x1 x2
      Y = y }

/// 原著と同じ隠れ層を持つネットワークを学習する。
///
/// Accord の `ActivationNetwork` は「活性化関数・入力数・各層のユニット数」を取る。
/// **層ごとに活性化関数を変えられない** ので、隠れ層も出力層もシグモイドになる。
/// Keras のように ReLU と softmax を混ぜることはできない。
///
/// Accord には Dropout も無い。原著が過学習を抑えるために入れている 2 か所は、
/// ここでは省くことになる。
///
/// **`NguyenWidrow(...).Randomize()` が必須である。** これを呼ばないと重みが
/// 初期化されず、何エポック回しても学習が進まない（常に同じクラスを答え、
/// 正解率 0.236 のまま）。Keras は層を作った時点で自動的に初期化するので、
/// 移植のときに見落としやすい。
let fit (data: Circle) (epochs: int) (seed: int) =
    // Accord は内部で乱数を使うので、種を固定して再現性を確保する
    Accord.Math.Random.Generator.Seed <- System.Nullable seed

    let network =
        ActivationNetwork(SigmoidFunction(), 2, hiddenUnits.[0], hiddenUnits.[1], 1)

    // 重みの初期化。これが無いと学習が始まらない
    NguyenWidrow(network).Randomize()

    let teacher = BackPropagationLearning(network)
    teacher.LearningRate <- LearningRate
    let outputs = data.Y |> Array.map (fun label -> [| float label |])

    for _ in 1..epochs do
        teacher.RunEpoch(data.X, outputs) |> ignore

    network

/// 重みを初期化せずに学習する。比較用。
///
/// `NguyenWidrow` を呼ばない場合にどうなるかを示すために残してある。
let fitWithoutInitialization (data: Circle) (epochs: int) (seed: int) =
    Accord.Math.Random.Generator.Seed <- System.Nullable seed

    let network =
        ActivationNetwork(SigmoidFunction(), 2, hiddenUnits.[0], hiddenUnits.[1], 1)

    let teacher = BackPropagationLearning(network)
    teacher.LearningRate <- LearningRate
    let outputs = data.Y |> Array.map (fun label -> [| float label |])

    for _ in 1..epochs do
        teacher.RunEpoch(data.X, outputs) |> ignore

    network

/// 1 点を予測する。出力が 0.5 以上なら 1
let predict (network: ActivationNetwork) (point: float[]) =
    if network.Compute(point).[0] >= 0.5 then 1 else 0

/// 学習データに対する正解率
let accuracy (network: ActivationNetwork) (data: Circle) =
    let correct =
        Array.map2 (fun point label -> predict network point = label) data.X data.Y
        |> Array.filter id
        |> Array.length

    float correct / float data.Size

/// NumPy の `np.arange` と同じく、終端を含まない等差数列を作る
let arange (start: float) (stop: float) (step: float) =
    let count = int (ceil ((stop - start) / step))
    Array.init count (fun i -> start + float i * step)

/// 原著の `plot_model` と同じ格子を作り、各点の予測クラスを返す
let decisionGrid (network: ActivationNetwork) (data: Circle) (step: float) =
    let xs = data.X |> Array.map (fun p -> p.[0])
    let ys = data.X |> Array.map (fun p -> p.[1])
    let xValues = arange (Array.min xs - 1.0) (Array.max xs + 1.0) step
    let yValues = arange (Array.min ys - 1.0) (Array.max ys + 1.0) step

    yValues
    |> Array.map (fun yValue -> xValues |> Array.map (fun xValue -> predict network [| xValue; yValue |]))

/// 各行で予測が切り替わった回数を返す。
///
/// 決定木なら軸に平行な境界なので、切り替わる位置は行によらず同じだった。
/// ニューラルネットワークは曲線を引けるので、行ごとに変わる。
let boundaryChangesPerRow (grid: int[][]) =
    grid
    |> Array.map (fun row -> row |> Array.pairwise |> Array.filter (fun (a, b) -> a <> b) |> Array.length)
    |> List.ofArray
